### PR TITLE
[charts] Refactor API using SIP-35

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [X] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Refactor of the Dashboard API using SIP-35

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@willbarrett @craig-rueda @nytai

### CODE REVIEW

File: superset/charts/dao.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+class ChartDAO(BaseDAO):
+    model_cls = Slice
+    base_filter = ChartFilter
+
+    @staticmethod
+    def get_dashboards_by_ids(dashboard_ids: List[int]) -> List[Dashboard]:
+        return db.session.query(Dashboard).filter(Dashboard.id.in_(dashboard_ids)).all()
+
+    @classmethod
+    def create(cls, properties: Dict, commit=True) -> Optional[Slice]:
```
Comment (craig-rueda) [inline superset/charts/dao.py]: `cls` param still here

File: superset/dao/base.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    @classmethod
+    def find_by_id(cls, model_id: int) -> Model:
+        data_model = SQLAInterface(cls.model_cls, db.session)
+        query = db.session.query(cls.model_cls)
+        query = cls.base_filter("id", data_model).apply(query, None)
+        return query.filter_by(id=model_id).one_or_none()
+
+    @classmethod
+    def create(cls, properties: Dict, commit=True) -> Optional[Model]:
```
Comment (craig-rueda) [inline superset/dao/base.py]: Do we need to take `cls` as a param?

File: superset/dao/base.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+)
+from superset.extensions import db
+
+
+class BaseDAO:
+    model_cls: Optional[Model] = None
+    base_filter: Optional[BaseFilter] = None
+
+    @classmethod
+    def find_by_id(cls, model_id: int) -> Model:
```
Comment (craig-rueda) [inline superset/dao/base.py]: It's unclear what `cls` is here. Ideally, we would just have a sig like `def find_by_id(id: int) -> Model`

File: superset/charts/dao.py (hunk 1 above)
Comment (craig-rueda) [reply superset/charts/dao.py]: This is a bit confusing, as a ChartDAO should simply CRUD on Slices. We shouldn't need, or want, to pass in a model class for each invocation. The ChartDAO should already know what its model class is (as I see you have declared in the class prop)

**Review by craig-rueda — CHANGES_REQUESTED** [review]

**Review by craig-rueda — APPROVED** [review]

**Comment by serenajiang** [comment]
@dpgaspar This PR broke "edit properties" for charts. `GET api/v1/chart/{chart_id}` seems to return  `404` for all charts. Is there a quick fix for this?

**Comment by serenajiang** [comment]
@craig-rueda for more eyes

**Comment by dpgaspar** [comment]
I'm getting errors on a fresh virtualenv using marshmallow > 3. Flask-AppBuilder setup.py sets the version constraints to "marshmallow>=2.18.0, <4.0.0". Pinned requirements.txt on superset and Flask-AppBuilder don't present any problems. Can you confirm?

Going to address this on FAB ASAP (pin marshmallow < 3) and release 2.3.1

**Comment by dpgaspar** [comment]
@serenajiang 
Yet my previous comment may be unrelated, since it would return 500 on an unpinned venv, not 404. Not being able to reproduce

**Comment by serenajiang** [comment]
I am reproducing this error with marshmallow `2.19.5`. We don't see the same problem with the dashboards refactor (#9315).

**Pull request merged**